google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java

```diff
@@ -16,15 +16,85 @@
+import com.google.cloud.tools.intellij.ui.GoogleCloudToolsIcons;
+import com.google.cloud.tools.intellij.util.GctBundle;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.ui.ComboBox;
+import com.intellij.openapi.ui.DialogWrapper;
+import com.intellij.openapi.ui.ValidationInfo;
+import com.intellij.ui.DocumentAdapter;
+import com.intellij.ui.ListCellRendererWrapper;
+import com.intellij.ui.TableSpeedSearch;
+import com.intellij.ui.components.JBTextField;
+import com.intellij.ui.table.JBTable;
+import git4idea.DialogManager;
+import java.awt.BorderLayout;
+import java.awt.event.ActionEvent;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+import javax.swing.AbstractAction;
+import javax.swing.Action;
+import javax.swing.JButton;
+import javax.swing.JComboBox;
+import javax.swing.JComponent;
+import javax.swing.JList;
+import javax.swing.JPanel;
+import javax.swing.JTable;
+import javax.swing.JTextField;
+import javax.swing.ListSelectionModel;
+import javax.swing.SwingUtilities;
+import javax.swing.event.DocumentEvent;
+import javax.swing.table.TableModel;
+import javax.swing.table.TableRowSorter;
+import org.jdesktop.swingx.sort.RowFilters.GeneralFilter;
+import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * Modal project and account selection dialog. Contains account drop-down with user list, table with
  * project list and simple filter. {@link ProjectSelector} calls {@link #showDialog(CloudProject)}.
+ * Calls {@link ProjectLoader#} to load projects for all known accounts and caches them.
  */
-// TODO(ivanporty) implementation in the following PRs
 public class ProjectSelectionDialog {
 
+  private JComboBox<CredentialedUser> accountComboBox;
+  private JButton addAccountButton;
+  private JTextField filterTextField;
+  private JTable projectListTable;
+  private JPanel centerPanel;
+
+  private ProjectSelectionDialogWrapper dialogWrapper;
+
+  private JPanel centerPanelWrapper;
+  private ProjectSelectorSignInPanel signInScreen = new ProjectSelectorSignInPanel();
+
+  private RefreshAction refreshAction;
+
+  private ProjectListTableModel projectListTableModel;
+
+  private ProjectLoader projectLoader = new ProjectLoader();
+
+  private Map<CredentialedUser, ListenableFuture<List<Project>>> runningProjectLoaderJobs =
+      Maps.newHashMap();
+  // last project name selection made by user for an account.
+  private Map<CredentialedUser, String> selectedProjectsByAccount = Maps.newHashMap();
```

Comment:
what do you think about combining selecteProjectsByAccount and cachedProjectList into a single mutlimap?

Reply:
specifically, guava's multimap

Reply:
you mean selected project is marked somehow in project list, i.e. we have Multimap<User,Project> and selected project is one of them?

Reply:
Never mind, it may have been a bad idea.